### PR TITLE
Use global variables to register the return_types namedtuples

### LINKED ISSUE #69221: torch.return_types: simplify implementation
## 🚀 Feature

Follow up to this thread: https://github.com/pytorch/pytorch/pull/66614#discussion_r745505413

The implementation of torch.return_types uses a static map. This unfortunately duplicates the memory needed for torch.return_types (but the total memory needed is not that much). We should try to get rid of the map by turning the entries in the map into globals that get set during initialization and used directly.

## Motivation

Decrease the amount of memory used by PyTorch; also Better Engineering

## Pitch

We should try to get rid of the map by turning the entries in the map into globals that get set during initialization and used directly.

## Alternatives

Use a constexpr map. This seems like overkill

## Additional context

n/a

cc @ezyang @bhosmer @smessmer @ljk53 @bdhirsh

### PR DESCRIPTION
**Pull request by andreigh**
Fixes #69221

@pytorchbot label "topic: not user facing"

### CODE REVIEW

**Comment by andreigh** [comment]
@zou3519 can you please take a look ? Thank you

**Comment by andreigh** [comment]
Fixed the lint and the CI error

**Comment by andreigh** [comment]
@pytorchbot label "suppress-api-compatibility-check"
@pytorchbot label "suppress-bc-linter"

Disabled BC Linter [which is detecting](https://github.com/pytorch/pytorch/actions/runs/5861630234/job/15892079506) the deletion of the "generate_return_type_definition_and_map_entry" function

File: tools/autograd/templates/python_fft_functions.cpp [hunk 1]
```diff
@@ -5,7 +5,7 @@
 #include "torch/csrc/DynamicTypes.h"
 #include "torch/csrc/Exceptions.h"
 #include "torch/csrc/autograd/python_fft_functions.h"
-#include "torch/csrc/autograd/python_return_types.h"
+#include "torch/csrc/autograd/generated/python_return_types.h"
```
Comment (zou3519) [inline tools/autograd/templates/python_fft_functions.cpp:8]: Could you send an example of the generated code? Perhaps in a gist?

Comment (andreigh) [reply tools/autograd/templates/python_fft_functions.cpp:8]: Here are the entire generated files content: https://gist.github.com/andreigh/b1b282c0ba40666a979457057fe94bb2

File: build.bzl [hunk 2]
```diff
@@ -255,6 +255,7 @@ GENERATED_CPP_CORE = [
 
 _GENERATED_AUTOGRAD_PYTHON_HEADERS = [
     "torch/csrc/autograd/generated/python_functions.h",
+    "torch/csrc/autograd/generated/python_return_types.h",
```
[the remark this answers is not in the document]
Comment (andreigh) [reply build.bzl:258]: I don't know either, I've just stepped in to work on some beginner issues still open.
The map was keeping extra pointers, but that's not a full duplicate of the PyTypeObject variables.

[the remark this answers is not in the document]
Comment (andreigh) [reply build.bzl:258]: Also this is definitely better engineering, the map was useless in the initial code.

[the remark this answers is not in the document]
Comment (zou3519) [reply build.bzl:258]: Thanks for your patience. I remembered enough to say yes, the problem is that the map is not necessary :)

File: tools/autograd/gen_python_functions.py [hunk 3]
```diff
@@ -735,9 +770,42 @@ def generate_return_type_definition_and_map_entry(
...
+        str, str
+    ] = {}  # map from unique name + field name lists to typedef name
+    declarations: List[str] = []  # function declaration to register the typedef
+
+    for overload in overloads:
+        fieldnames = namedtuple_fieldnames(overload.function.func.returns)
+        if not fieldnames:
+            continue
+
+        name = cpp.name(overload.function.func)  # use @with_native_function?
```
Comment (zou3519) [inline tools/autograd/gen_python_functions.py:797]: What's the question here?

**Review by zou3519 — APPROVED** [review]
This LGTM. Let's wait for the CI to run

File: tools/autograd/gen_python_functions.py (hunk 3 above)
Comment (andreigh) [reply tools/autograd/gen_python_functions.py:797]: I have no idea, just copied from the above function - I can delete the comment

**Comment by andreigh** [comment]
[Some of the failed tests](https://github.com/pytorch/pytorch/actions/runs/5869413407/job/16113184547) succeed locally (rebased at main 2 days ago):

    (pytorch) C:\work\Repos\pytorch\test>python test_ops_jit.py -k test_variant_consistency_jit_addmm_decomposed_cpu_complex64
    .
    ----------------------------------------------------------------------
    Ran 1 test in 11.745s
    
    OK

While [the rest of the tests](https://github.com/pytorch/pytorch/actions/runs/5869413407/job/16113184736) fail because the Pytorch being tested is not compiled with LAPACK:

    RuntimeError: Calling torch.linalg.lu_factor on a CPU tensor requires compiling PyTorch with LAPACK. Please use PyTorch built with LAPACK support.

Is it possible to rebase at viable/strict and then re-run CI ? Thank you !

**Comment by andreigh** [comment]
I've compiled locally with LAPACK and the tests (test_dtypes__refs_linalg_svd_cp) ran successfully.

Is there any way you can trigger a rebase ?

Thank you !

**Comment by zou3519** [comment]
@andreigh usually I'll just rebase locally and then re-push to the branch. Something like `git pull --rebase upstream viable/strict` and `git push --force`

**Comment by andreigh** [comment]
[CI is green](https://github.com/pytorch/pytorch/actions/runs/6030483038/job/16363076508)

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change, but it is failing internal build' -c ghfirst

Here is the error FYI:

```
stderr: Traceback (most recent call last):
  File "<string>", line 49, in <module>
  File "<string>", line 47, in __run
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/gen/7be98315/xplat/caffe2/tools/generate_code_bin#link-tree/tools/setup_helpers/generate_code.py", line 236, in <module>
    main()
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/gen/7be98315/xplat/caffe2/tools/generate_code_bin#link-tree/tools/setup_helpers/generate_code.py", line 187, in main
    generate_code(
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/gen/7be98315/xplat/caffe2/tools/generate_code_bin#link-tree/tools/setup_helpers/generate_code.py", line 46, in generate_code
    gen_autograd_python(
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/gen/7be98315/xplat/caffe2/tools/generate_code_bin#link-tree/tools/autograd/gen_autograd.py", line 116, in gen_autograd_python
    gen_python_functions.gen(
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/gen/7be98315/xplat/caffe2/tools/generate_code_bin#link-tree/tools/autograd/gen_python_functions.py", line 362, in gen
    create_python_return_type_bindings_header(
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/gen/7be98315/xplat/caffe2/tools/generate_code_bin#link-tree/tools/autograd/gen_python_functions.py", line 494, in create_python_return_type_bindings_header
    fm.write_with_template(
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/gen/7be98315/xplat/caffe2/tools/generate_code_bin#link-tree/torchgen/utils.py", line 179, in write_with_template
    substitute_out = self.substitute_with_template(
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/gen/7be98315/xplat/caffe2/tools/generate_code_bin#link-tree/torchgen/utils.py", line 162, in substitute_with_template
    template = _read_template(template_path)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/gen/7be98315/xplat/caffe2/tools/generate_code_bin#link-tree/torchgen/utils.py", line 113, in _read_template
    return CodeTemplate.from_file(template_fn)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/gen/7be98315/xplat/caffe2/tools/generate_code_bin#link-tree/torchgen/code_template.py", line 23, in from_file
    with open(filename) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/gen/7be98315/xplat/caffe2/tools/generate_code_bin#link-tree/tools/autograd/templates/python_return_types.h'

    When running <genrule>.
    When building rule //xplat/caffe2:gen_aten_libtorch (ovr_config//platform/linux:x86_64-clang-opt-platform010).
```

I will add `ciflow/periodic` to run buck build and test to see if it could catch this failure on OSS

**Comment by andreigh** [comment]
> @pytorchbot revert -m 'Sorry for reverting your change, but it is failing internal build' -c ghfirst

Do you know which build files for the internal build use ? Is it buckbuild.bzl ?

**Comment by andreigh** [comment]
@huydhn can you please try buck build with #108832 ? I updated some buck build files. Thank you !

**Pull request closed without merging**